### PR TITLE
Add image uploading

### PR DESCRIPTION
**Pull request by heyitsmeuralex**
This PR merges into componentize, not master.

* `` - you can now include images in your markdown ✔️ ⬇️ 
* pasting inside the message editor will upload an image (you don't have to be admin anymore! this should be a configuration option per-server, I guess) and send a message including it :tada: 📋

### CODE REVIEW

File: api.js [hunk 1]
```diff
@@ -429,17 +429,17 @@ module.exports = async function attachAPI(app, {wss, db}) {
...
 
         req[middleware.vars].path += '/' + filename
-
         cb(null, filename)
       }
     })
   })
   const uploadSingleImage = upload.single('image')
 
+  // TODO: delete old images
```
Comment (towerofnix) [inline api.js]: This should be a server option. Also, custom emoji images should *not* be deleted, which will be something to consider.

File: site/js/components/message-editor.js [hunk 2]
```diff
@@ -37,10 +37,51 @@ const component = (state, emit) => {
...
+
+      // send a message with the image in it
+      await api.post(state, 'send-message', {
+        text: `![](${state.secure ? 'https' : 'http'}://${state.params.host}${path})`,
+        channelID: state.params.channel,
+        sessionID: state.session.id,
+      })
+
+      progressBar.style.width = '100%'
+      await api.sleep(500)
```
Comment (towerofnix) [inline site/js/components/message-editor.js]: Why the delay? I imagine it visually looks a bit better?

File: site/js/components/message-group.js [hunk 3]
```diff
@@ -4,6 +4,28 @@ const raw = require('choo/html/raw')
...
+  meta({ alt, src })
+
+  return read() === ')'
+}
+
+mrk.htmlify.image = ({ metadata: { alt, src } }) =>
+  `<a href='${mrk.escapeHTML(src)}' target='_blank' class='image'>
+    <img src='${mrk.escapeHTML(src)}' alt='${mrk.escapeHTML(alt)}'s/>
+  </a>`
+
```
Comment (towerofnix) [inline site/js/components/message-group.js]: :tada:

File: site/js/components/messages.js [hunk 4]
```diff
@@ -208,6 +208,16 @@ const store = (state, emitter) => {
...
           block: 'end',
         })
+
+        let img
+        if (img = el.querySelector('.image:last-of-type img')) {
+          // if the message has an image in it, wait for the image to load,
+          // then scroll down to it
+          img.addEventListener('load', () => {
+            setTimeout(() =>
+              state.messages.scrollToMsg(message), 25)
```
Comment (towerofnix) [inline site/js/components/messages.js]: Code style nitpick - should be this:

```js
setTimeout(() => {
  state.messages.scrollToMsg(message)
}, 25)
```

..or this:

```js
setTimeout(() => state.messages.scrolLToMsg(message), 25)
```

File: site/js/util/api.js [hunk 5]
```diff
@@ -49,5 +49,14 @@ module.exports = {
     })
   },
 
+  postRaw(state, path, body = '') {
+    if (!body) return Promise.reject(new Error('Body not provided'))
```
Comment (towerofnix) [inline site/js/util/api.js]: Does it make sense to provide a default value of `''` when that default value won't actually work? I'm thinking no default value at all.

File: site/js/components/message-editor.js (hunk 2 above)
Comment (heyitsmeuralex) [reply site/js/components/message-editor.js]: Yep.

File: site/js/util/api.js (hunk 5 above)
Comment (heyitsmeuralex) [reply site/js/util/api.js]: Oh yeah.

File: api.js (hunk 1 above)
Comment (heyitsmeuralex) [reply api.js]: ...should this be added later?

**Comment by heyitsmeuralex** [comment]
I'll create an issue for deleting stale uploads.

**Pull request merged**